src/library/scala/concurrent/impl/Promise.scala

```diff
@@ -7,400 +7,457 @@
 
 package scala.concurrent.impl
-
-import scala.concurrent.{ ExecutionContext, CanAwait, OnCompleteRunnable, TimeoutException, ExecutionException }
-import scala.concurrent.Future.InternalCallbackExecutor
-import scala.concurrent.duration.{ Duration, FiniteDuration }
-import scala.annotation.tailrec
-import scala.util.control.NonFatal
+import scala.concurrent.{ExecutionContext, CanAwait, TimeoutException, ExecutionException, Future, Batchable }
+import Future.InternalCallbackExecutor
+import scala.concurrent.duration.Duration
+import scala.annotation.{ tailrec, unchecked, switch }
+import scala.util.control.{ NonFatal, ControlThrowable }
 import scala.util.{ Try, Success, Failure }
-
+import scala.runtime.NonLocalReturnControl
 import java.util.concurrent.locks.AbstractQueuedSynchronizer
 import java.util.concurrent.atomic.AtomicReference
+import java.util.Objects.requireNonNull
+
+/**
+  * Latch used to implement waiting on a DefaultPromise's result.
+  *
+  * Inspired by: http://gee.cs.oswego.edu/cgi-bin/viewcvs.cgi/jsr166/src/main/java/util/concurrent/locks/AbstractQueuedSynchronizer.java
+  * Written by Doug Lea with assistance from members of JCP JSR-166
+  * Expert Group and released to the public domain, as explained at
+  * http://creativecommons.org/publicdomain/zero/1.0/
+  */
+private[impl] final class CompletionLatch[T] extends AbstractQueuedSynchronizer with (Try[T] => Unit) {
+  //@volatie not needed since we use acquire/release
+  /*@volatile*/ private[this] var _result: Try[T] = null
+  final def result: Try[T] = _result
+  override protected def tryAcquireShared(ignored: Int): Int = if (getState != 0) 1 else -1
+  override protected def tryReleaseShared(ignore: Int): Boolean = {
+    setState(1)
+    true
+  }
+  override def apply(value: Try[T]): Unit = {
+    _result = value // This line MUST go before releaseShared
+    releaseShared(1)
+  }
+}
 
-private[concurrent] trait Promise[T] extends scala.concurrent.Promise[T] with scala.concurrent.Future[T] {
-  def future: this.type = this
-
-  import scala.concurrent.Future
-  import scala.concurrent.impl.Promise.DefaultPromise
+private[concurrent] final object Promise {
+    /**
+     * Link represents a completion dependency between 2 DefaultPromises.
+     * As the DefaultPromise referred to by a Link can itself be linked to another promise
+     * `relink` traverses such chains and compresses them so that the link always points
+     * to the root of the dependency chain.
+     *
+     * In order to conserve memory, the owner of a Link (a DefaultrPomise) is not stored
+     * on the Link, but is instead passed in as a parameter to the operation(s).
+     *
+     * If when compressing a chain of Links it is discovered that the root has been completed,
+     * the `owner`'s value is completed with that value, and the Link chain is discarded.
+     **/
+    private final class Link[T](to: DefaultPromise[T]) extends AtomicReference[DefaultPromise[T]](to) {
```

Comment:
Because Transformation both IS-A DefaultPromise and a Callback. This allows us to more cheaply have Transformations with a single callback, which tends to be the main case for transformations.

Reply:
Are talking about the fact that because a Transformation is both a DefaultPromise and a Callback then that means you can't use simple instanceOf checks to determine whether it should be treated as a callback or as a linked-to promise?

Instead of using pure instanceOf checks, could you query the Transformation's state to work out whether it needs to be treated as linked or as awaiting a callback?

Reply:
Yes.

The problem is that it being a link is a property of the current promise and not its callback/Transformation. And introducing a new field would make it impossible to do an atomic swap.

Reply:
Sorry I haven't replied ... I'm still digesting this. Can we look at a concrete example? I'm thinking about typical 'flatMap' where we have three promises:

```scala
val p1 = Promise[Int]() // initial promise that we'll flatMap
val p2 = Promise[Int]() // will yield result of the flatMap

val p3 = p1.flatMap(_ => p2) // Transformation of type 'flatMap'
// p3 is a Transformation; there is now a callback from p1 to p3

p1.success(1)
// calls the p3 Transformation callback; creates a 'link' from p3 to p2
// Question: could p3's 'link' just be a direct reference to p2 instead of needing a special Link object? I can't see why not?

p2.success(1)
// Completes p2 and any linked promises (e.g. p3)
```

Reply:
@richdougherty My implementation had a bug in that case, and there were no tests that verified it. I have now both fixed that issue AND added a regression test.

For your question: In this case it could refer to the DefaultPromise instance (technically) because DefaultPromise does not implement Callback. But, in the case where you have `val p3 = p1.flatMap(_ => p2.future.map(identity)) it would be a Transformation, which implements Callback, so it would need to be in Link. Or?

If you run the benches on your machine against current 2.13.x vs my proposed implementation here, what does those results look like?

Reply:
@viktorklang, I see what you mean. We can use `instanceof Link` to determine if it's a link relationship vs a callback relationship. The other option, which might avoid creating extra `Link` objects could be to use an `AtomicStampedReference` so that we have both a reference and an `int` part that we can update atomically. We could use the `int` part to differentiate whether a promise is a callback or a link. That might save some extra object creation.

Reply:
@richdougherty That's an interesting idea! Unfortunately that won't really gain us anything from an allocation point-of-view: "Implementation note: This implementation maintains stamped references by creating internal objects representing "boxed" [reference, integer] pairs." - https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/atomic/AtomicStampedReference.html

Reply:
Damn! I read that as meaning they used native code to allocate a boxed pair inside the already-allocated `AtomicStampedReference` object's memory, but I just looked at the source see I was mistaken.

The only other idea I have is to use an `Atomic*FieldUpdater` to store an extra bit of information, but I don't think those really work in Scala (due to how companion objects are generated).

So, I'm happy for you to go ahead with your Link/Transformation encoding.